File: leetcode-solution/leetcode28-lintcode13-strStr/src/Solution.java
/**
 * Created by Felicity on 2017/3/15.
 * 这是比较笨的一种方法
 * 用source的长度减去target的长度，然后再用target去和source挨个比对，如果target的第一个字母对上了就比对第二个，第二个对上了就比对第三个，以此类推，
 * 如果不对，把target往后移一位继续重复刚才的比对
 */
class Solution {
    /**
     * Returns a index to the first occurrence of target in source,
     * or -1  if target is not part of source.
     * @param source string to be scanned.
     * @param target string containing the sequence of characters to match.
     */
    public int strStr(String source, String target) {
        // write your code here 0;
        if (source == null || target == null) {
            return -1;
        }
        for (int i = 0; i < source.length() - target.length() + 1; i++) { //这里是用source的长度减去target的长度
            int j = 0;
            for (j = 0; j < target.length(); j++) { //用target里面的每一个字母去对比
                if (source.charAt(i + j) != target.charAt(j)) { //一旦有不一样的字母，立刻停止
                    break;//break 是结束程序不往下走了，一般用在循环中，终止循环的，continue在循环中表示跳出本次循环，直接进行下一次循环
                }
            }
            if (j == target.length()) { //如果从source的i处，数了j个字母与target的长度相同（也就是和target完全一样），那么就返回那个i值
                return i;
            }
        }
        return -1;
    }
}

File: leetcode-solution/leetcode28-lintcode13-strStr/Solution.py
# Python的类的方法和普通的函数有一个很明显的区别，在类的方法必须有个额外的第一个参数(self)，但在调用这个方法
# 的时候不必为这个参数赋值(显胜于隐的引发）。Python的类的方法的这个特别的参数指代的是对象本身，而按照Python
# 的惯例，它用self来表示。（当然我们也可以用其他任何名称来代替，只是规范和标准在那建议我们一致使用self）

class Solution:
    """
    @param: source: source string to be scanned.
    @param: target: target string containing the sequence of characters to match
    @return: a index to the first occurrence of target in source, or -1  if target is not part of source.
    """
    def strStr(self, source, target):
        # write your code here
        if source is None or target is None:
            return -1
        return source.find(target)
    # Python find() 方法检测字符串中是否包含子字符串 str ，如果指定 beg（开始）
    # 和 end（结束） 范围，则检查是否包含在指定范围内，如果包含子字符串返回开始
    # 的索引值，否则返回-1。



# 以下的方法是不用python内置的find()的方法
# class Solution:
#     """
#     @param: source: source string to be scanned.
#     @param: target: target string containing the sequence of characters to match
#     @return: a index to the first occurrence of target in source, or -1  if target is not part of source.
#     """
#     def strStr(self, source, target):
#         # write your code here
#         if source is None or target is None:
#             return -1
#         for i in range (len(source) - len(target) + 1):
#             if source[i:i + len(target)] == target:
#                 return i
#         return -1
